src/test-apps/happy/lib/WeaveDeviceManager.py

```diff
@@ -446,6 +470,65 @@ def testWdmClientDataSinkRefreshGetDataRefresh(testObject):
     testObject.closeWdmClient()
     print("testWdmClientDataSinkRefreshGetDataRefresh completes")
 
+def testWdmClientDataSinkRefreshDataWithEvents(testObject):
+    # Sleep 5 second for enough events.
+    testObject.createWdmClient()
+    time.sleep(5)
+    # We need to set node Id to get everything works
+    testObject.setNodeId(-2)
+    # We needs a data sink anyway
+    localeSettingsTrait = testObject.newDataSink(20, 0, b"/")
+    testObject.setEventFetchingTimeout(1)
+    testObject.refreshData(withEvents=True)
+    data1 = testObject.getEvents()
+    # Sleep 5 second to wait for more events generated by mock-device
+    time.sleep(5)
+    testObject.refreshData(withEvents=True)
+    data2 = testObject.getEvents()
+
+    jsondata1 = json.loads(data1.decode("utf-8"))
+    jsondata2 = json.loads(data2.decode("utf-8"))
+    # Mockdevice should generate events
+    if (len(jsondata1) == 0) or (len(jsondata2) == 0):
+        raise ValueError("testWdmClientDataSinkRefreshDataWithEvents  fails")
+    # Event should not overlap
+    if jsondata2[0]["Id"] <= jsondata1[-1]["Id"]:
+        raise ValueError("testWdmClientDataSinkRefreshDataWithEvents  fails")
+    # Sleep 5 second to wait for more events generated by mock-device
+    time.sleep(5)
+    testObject.refreshData()
+    val = testObject.getEvents()
+    # Not equals to []
+    if val.decode("utf-8") != "[]":
+        raise ValueError("testWdmClientDataSinkRefreshDataWithEvents  fails")
+
+    testObject.closeWdmClient()
+    print("testWdmClientDataSinkRefreshDataWithEvents completes")
+
+def testWdmClientDataSinkRefreshDataWithLargeEventList(testObject):
+    # Mockdevice should generate event at 1 per second.
+    testObject.createWdmClient()
+    # Sleep 12 second for enough events, connection timeout is 15s
+    time.sleep(12)
+    # We need to set node Id to get everything works
+    testObject.setNodeId(-2)
+    # We needs a data sink anyway
+    localeSettingsTrait = testObject.newDataSink(20, 0, b"/")
```

Comment:
two tests; one is with trait, one is without trait
please update the check of non-empty list in PathList::Parser::CheckSchemaValidity